app/actions.py

```diff
@@ -92,30 +92,170 @@ def performDeleteRange(self, upperRow, upperCol, lowerRow, lowerCol):
         self.getText(upperRow, upperCol, lowerRow, lowerCol)))
       self.redo()
 
+  def dataToBookmark(self):
+    """
+    Grabs all the cursor data and returns a bookmark.
+
+    Args:
+      None.
+
+    Returns:
+      A bookmark in the form of (bookmarkRange, bookmarkData).
+      bookmarkRange is an ordered tuple in which its elements
+      are the rows that the bookmark affects.
+      bookmarkData is a dictionary that contains the cursor data.
+    """
+    cursor = (self.view.cursorRow, self.view.cursorCol)
+    pen = (self.penRow, self.penCol)
+    marker = (self.markerRow, self.markerCol)
+    selectionMode = self.selectionMode
+    bookmarkData = {
+      'cursor': cursor,
+      'marker': marker,
+      'pen': pen,
+      'selectionMode': selectionMode,
+    }
+    upperRow, _, lowerRow, _ = self.startAndEnd()
+    bookmarkRange = tuple(row for row in range(upperRow, lowerRow + 1))
+    return (bookmarkRange, bookmarkData)
+
+  def bookmarksOverlap(self, bookmarkRange1, bookmarkRange2):
+    """
+    Returns whether the two sorted bookmark ranges overlap.
+    
+    Args:
+      bookmarkRange1 (tuple): a sorted tuple of row numbers.
+      bookmarkRange2 (tuple): a sorted tuple of row numbers.
+
+    Returns:
+      True if the ranges overlap. Otherwise, returns False.
+    """
+    if (bookmarkRange1[-1] >= bookmarkRange2[0] and
+        bookmarkRange1[0] <= bookmarkRange2[-1]):
+      return True
+    else:
+      return False
+
   def bookmarkAdd(self):
-    app.bookmarks.add(self.fullPath, self.penRow, self.penCol, 0,
-        app.selectable.kSelectionNone)
+    """
+    Adds a bookmark at the cursor's location. If multiple lines are 
+    selected, all existing bookmarks in those lines are overwritten 
+    with the new bookmark.
 
-  def bookmarkGoto(self):
+    Args:
+      None.
+
+    Returns:
+      None.
+    """
+    newBookmark = self.dataToBookmark()
+    self.bookmarkRemove()
+    bisect.insort(self.bookmarks, newBookmark)
+
+  def bookmarkGoto(self, bookmark):
+    """
+    Goes to the bookmark that is passed in.
+
+    Args:
+      bookmark (tuple): contains bookmarkRange and bookmarkData. More info can
+        be found in the dataToBookmark function.
+
+    Returns:
+      None.
+    """
     app.log.debug()
-    bookmark = app.bookmarks.get(self.view.bookmarkIndex)
-    if bookmark:
-      self.selectText(bookmark['row'], bookmark['col'], bookmark['length'],
-          bookmark['mode'])
+    bookmarkRange, bookmarkData = bookmark
+    cursorRow, cursorCol = bookmarkData['cursor']
+    penRow, penCol = bookmarkData['pen']
+    markerRow, markerCol = bookmarkData['marker']
+    selectionMode = bookmarkData['selectionMode']
+    self.__skipCursorScroll = True
```

Comment:
This might be something from another PR. Let's separate it from this PR. i.e. delete line 172.

Reply:
woops :D got too excited working on it